### DISCUSSION TITLE
Understanding the Relationship between number Samples, Sweeps, Chains, and Model Features in the Aziz Potential Code Example

### DISCUSSION THREAD

**Discussion by MrAAafa99** [netket/netket#1524] [Q&A] [answered]
Hello! 
I have a question according to the "Interacting system with continuous degrees of freedom" section and the provided code example for Aziz potential. 

        N = 10
        d = 0.3 # 1/Angstrom
        rm = 2.9673 # Angstrom
        L = N / (0.3 * rm)
        hilb = nk.hilbert.Particle(N=N, L=L, pbc=True)
        sab = nk.sampler.MetropolisGaussian(hilb, sigma=0.05, n_chains=16, n_sweeps=32)

        def minimum_distance(x, sdim):
          """Computes distances between particles using mimimum image convention"""
             n_particles = x.shape[0] // sdim
             x = x.reshape(-1, sdim)
             distances = (-x[jnp.newaxis, :, :] + x[:, jnp.newaxis, :])[
             jnp.triu_indices(n_particles, 1)
             ]
           distances = jnp.remainder(distances + L / 2.0, L) - L / 2.0
           return jnp.linalg.norm(distances, axis=1)

        def potential(x, sdim):
         """Compute Aziz potential for single sample x"""
          eps = 7.846373
          A = 0.544850 * 10 ** 6
          alpha = 13.353384
          c6 = 1.37332412
          c8 = 0.4253785
          c10 = 0.178100
          D= 1.241314

          dis = minimum_distance(x, sdim)
          return jnp.sum(
          eps
          * (
          A * jnp.exp(-alpha * dis)
          - (c6 / dis ** 6 + c8 / dis ** 8 + c10 / dis ** 10)
          * jnp.where(dis < D, jnp.exp(-((D / dis - 1) ** 2)), 1.0)
          )
          )
        ekin = nk.operator.KineticEnergy(hilb, mass=1.0)
        pot = nk.operator.PotentialEnergy(hilb, lambda x: potential(x, 1))
        ha = ekin + pot
        model = nk.models.DeepSetRelDistance(
        hilbert=hilb,
        cusp_exponent=5,
        layers_phi=2,
        layers_rho=3,
        features_phi=(16, 16),
        features_rho=(16, 16, 1),
        )
        vs = nk.vqs.MCState(sab, model, n_samples=4096, n_discard_per_chain=128)

        op = nk.optimizer.Sgd(0.001)
        sr = nk.optimizer.SR(diag_shift=0.01)

        gs = nk.VMC(ha, op, sab, variational_state=vs, preconditioner=sr)
        gs.run(n_iter=1000, out="Helium_10_1d")

       
In the code example, the total number of samples ( in "nk.vqs.MCState") is set to 4096=16**3. First of all, Why have you chosen this number as the number of samples? Does this number have any relationship with the number of sweeps or chains defined in the Metropolis sampler? If yes, does that mean I need to change those parameters when modifying the total number of samples? Alternatively, does the number of samples have any relationship with the features_phi and features_rho?

**Comment by gpescia** [comment, marked as answer]
Hi @MrAAafa99 
no the number of samples, sweeps, chains etc. is just a choice made to have reasonable precision on the energy while not taking too long for an optimization step. You can change them. The same holds for learning rate, diagonal shift, number of features in the network etc.
Hope that helps.

**Reply by MrAAafa99** [reply]
Thank you very much for your response. 
So why is "n_samples" equal to "n_sweeps * n_discard_per_chain" in this code? If I want to double, for example, "n_samples", should I also make these changes in "n_sweeps" and " n_discard_per_chain" in a way that preserves the relationship "n_samples=n_sweeps * n_discard_per_chain" as well?  Does setting these choices equal to powers of two improve convergence?  Or does determining these values somehow relate to Hilbert space?

**Reply by PhilipVinc** [reply]
`n_sweeps` is the subsampling factor of the Markov chain. Higher values lead to less correlated samples. Intuitively, larger moves (larger values of `sigma=0.05`) will lead to less correlated samples but possibly higher rejection rate, while smaller moves (smaller sigma) will lead to more correlated samples that require subsampling more aggressively by increasing `n_sweep`.

It's hard to give a general recipe that works, but what we do is check the autocorrelation time of the sampled chains (it is `tau_corr` in the expectation values of the energy and other observables) and try to play with the sampling parameters in order to reduce it to an acceptable value. 

`n_discard_per_chain` specifies how many effective samples (subsampled samples) are discarded from the chain every time parameters are changed in order to re-termalize the chain. 
This is also known as `warm up phase` in the literature.
If you subsample aggressively you probably can get away with a lower number of discarded samples (or warmup), but again it's hard to give a general recipe.
Also, if you have lots of short chains you probably need to increase this value.

**Reply by MrAAafa99** [reply]
@PhilipVinc, @gpescia 

I am very grateful for your response and the time you have dedicated. Thank you so much.
